### PR TITLE
Fall back to distributions without hashes in resolver

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This represents a change to `--require-hashes` in the event that we don't find a matching hash from the registry. The behavior in this PR is closer to pip's.

Prior to this PR, if a distribution had no reported hash, or only mismatched hashes, we would mark it as incompatible. Now, we mark it as compatible, but we use the hash-agreement as part of the ordering, such that we prefer any distribution with a matching hash, then any distribution with no hash, then any distribution with a mismatched hash.

As a result, if an index reports incorrect hashes, but the user provides the correct one, resolution now succeeds, where it would've failed.

Similarly, if an index omits hashes altogether, but the user provides the correct one, resolution now succeeds, where it would've failed.

If we end up picking a distribution whose hash ultimately doesn't match, we'll reject it later, after resolution.

### CODE REVIEW

File: crates/uv/tests/pip_sync.rs [hunk 1]
```diff
@@ -4159,12 +4286,20 @@ fn require_hashes_find_links_invalid_hash() -> Result<()> {
...
+    Resolved 1 package in [TIME]
+    error: Failed to download distributions
+      Caused by: Failed to fetch wheel: example-a-961b4c22==1.0.0
+      Caused by: Hash mismatch for example-a-961b4c22==1.0.0
+
+    Expected:
+      sha256:123
+
+    Computed:
+      sha256:5d69f0b590514103234f0c3526563856f04d044d8d0ea1073a843ae429b3187e
```
Comment (charliermarsh) [inline crates/uv/tests/pip_sync.rs:4302]: These aren't as nice-looking, but keep in mind that for `--require-hashes`, the user has to pin _all_ dependencies upfront anyway. There's no backtracking or anything.

Comment (zanieb) [reply crates/uv/tests/pip_sync.rs:4302]: They're pretty ugly but 🤷‍♂️ we could address later if we want.

**Review by zanieb — APPROVED** [review]
Yeah this UX seems much better.

File: crates/uv/tests/pip_sync.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/uv/tests/pip_sync.rs:4302]: Absolutely DESTROYING my error messages

**Pull request merged**